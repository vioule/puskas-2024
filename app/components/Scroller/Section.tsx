import { useInView } from "framer-motion";
import { useRef, useEffect, useContext } from "react";
import { AppContext, DataType } from "../DataProvider";

interface SectionProps {
  data: DataType;
}

export default function Section({ data }: SectionProps) {
  const ref = useRef(null);
  const { setData, lightbox, setLightbox } = useContext(AppContext);
  const timer = useRef<number | undefined>(undefined);
  const isInView = useInView(ref, { amount: 0.7, once: false });
  const color = isInView ? "text-white" : "text-gold";
  const cursor = lightbox.preview ? "hover:cursor-pointer" : "";

  const handleOnClick = () => {
    if (lightbox.preview) {
      setLightbox({
        preview: true,
        poster: true,
      });
    }
  };

  useEffect(() => {
    if (!lightbox.poster) {
      if (isInView) {
        timer.current = window.setTimeout(() => {
          setData(data);
          setLightbox({
            preview: true,
            poster: false,
          });
        }, 400);
      } else {
        window.clearTimeout(timer.current);
        setLightbox({
          preview: false,
          poster: false,
        });
      }
    }
    return () => clearTimeout(timer.current);
  }, [isInView, setData, data, setLightbox, lightbox.poster]);
  return (
    <div
      className={`flex justify-center items-center w-full h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] max-w-[800px] transition-[color] duration-1000 ease-linear ${color} ${cursor}`}
      ref={ref}
      onClick={handleOnClick}
      data-testid="section"
    >
      <span>{data.year}</span>
    </div>
  );
}
