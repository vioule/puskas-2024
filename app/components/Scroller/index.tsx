"use client";
import data from "@/app/data.json";
import Section from "./Section";
import { motion, MotionValue } from "framer-motion";

interface ScrollerProps {
  skewY: MotionValue<number>;
}

export default function Scroller({ skewY }: ScrollerProps) {
  return (
    <motion.div
      className="relative w-full mt-[100vh] border-t-8 border-white z-20"
      style={{ skewY }}
      data-testid="container"
    >
      {data.map((x) => {
        return <Section key={x.year} year={x.year} />;
      })}
    </motion.div>
  );
}
