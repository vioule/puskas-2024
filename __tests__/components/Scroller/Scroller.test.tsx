import { render } from "@testing-library/react";
import Scroller from "@/app/components/Scroller";
import { motionValue } from "framer-motion";

test("render Scroller unchanged", async () => {
  global.IntersectionObserver = jest.fn().mockReturnValue({
    observe: () => null,
    unobserve: () => null,
    disconnect: () => null,
  });
  const { container } = render(<Scroller skewY={motionValue(10)} />);
  expect(container).toMatchSnapshot();
});
