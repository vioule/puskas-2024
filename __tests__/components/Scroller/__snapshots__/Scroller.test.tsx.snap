// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Scroller unchanged 1`] = `
<div>
  <div
    class="relative w-full mt-[100vh] border-t-8 border-white z-20"
    data-testid="container"
    style="transform: skewY(10deg) translateZ(0);"
  >
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2009
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2010
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2011
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2012
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2013
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2014
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2015
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2016
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2017
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2018
      </span>
    </div>
    <div
      class="flex justify-center items-center h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] transition-[color] duration-1000 ease-linear text-gold"
    >
      <span>
        2019
      </span>
    </div>
  </div>
</div>
`;
